Update to latest upstream develop and apply fixes proposed upstream

JoeHut opened the pull request:
This patch brings in fixes from upstream and some additional memory leak fixes. This branch was created by:
* reverting the changes that have not been accepted upstream yet, but are present in our branch (asynchronous discovers)
* rebasing upstream develop on our branch and fast-forward merging it (effectively bringing in the upstream patches)
* applying the patches proposed here: https://github.com/hbldh/bleak/pull/154

arthur-proglove committed:
Do not start a scan if the timeout is <= 0

dlech committed:
backends/corebluetooth: add uuids metadata

This adds the received list of service UUIDs (if any)
to the BLEDevice metadata dictionary. UUIDs are
stored as list of lowercase string to match other
platforms.

dlech committed:
backends/corebluetooth: store manufacturer data as bytes

This changes the BLEDevice.metadata['manufacturer_data'] value
to be stored as a bytes object instead of as a hex
string. This is done so that it has the same format
on all platforms.

dlech committed:
backends/corebluetooth: use get() method

This uses the dictionary get() method to simplify the code.

dlech committed:
backends/corebluetooth: handle uuid.UUID arguments

There are several methods that can take a string or uuid.UUID as the
_uuid parameter. However, the get_appropriate_uuid method expects a
string, so be sure to convert it first. This is similar
to how the same methods are handled in other backends.

Commit:
feat(bluez): add support for asynchronous scanning of devices

This patch allows asynchronous discovery of nearby BLE devices. While
the blocking discover() function with a timeout is still available, an
additional class provides support for starting and stopping the
discovery session individually with a callback on each device change.

Commit:
fix(discovery): fix synchronous discover function

The discover function was not properly updated during development and
used an old signature. Update it to use the async discovery function
internally.

dlech committed:
bluez: add comment about version check

This version check gets questioned somewhat frequently. Each time,
it is found to be correct. So, let's add a comment with a detailed
explanation about what is going on.

Also add a check for BlueZ < 5.46 when attempting to use "Write without
response" to give a more useful error message.

hbldh committed:
Documentation improvements

Improved documentation for write_gatt_char
Moved #138 comment for code to docstring and to proper documentation.
Documented the new event loop problem detailed in #93 and #125

hbldh committed:
Improved error handling for Windows notify

Relates to #135

hbldh committed:
Improved type hints and handling of str/uuid

Fixes #127 for Windows and BlueZ backends.

Commit:
TypeError as dict was set to bool

hbldh committed:
Further improved type hints and handling of str/uuid

Improves #127 for Windows and BlueZ backends.

hbldh committed:
Minor fix for dotnet event loop handlers

hbldh committed:
Fixes #114.

hbldh committed:
Fixes #87.

hbldh committed:
Improved documentation on macOS differences.

Related to #140.

hbldh committed:
Abstract base class for new Scanner object.

Relates to #134, #131 and #128

hbldh committed:
Dotnet scanner object, first draft.

mickael9 committed:
backends/bluezdbus: use an a new reactor for each client

This makes things much cleaner :
- The default asyncio event loop is not modified
- No default twisted reactor is installed
- The reactor will use the provided event loop

Commit:
macos advertisementData fixes and more

Fixes the following issues on MacOS backend:

- advertisementData lost as multiple calls to
  centralManager_didDiscoverPeripheral_advertisementData_RSSI_
  for same device was not handled.

- Devices never removed only added while scanning and long running
  applications will get incorrect scan result.

- Scan was never stoped if timeout was 0 or None.

hbldh committed:
Residual changes for #143.

This tries to ensure that reactor is closed and deleted after client disconnects, to fix #130

hbldh committed:
Fixes for incorrectly checked in example files

hbldh committed:
Scanners API and improved docs

Implemented .NET Scanner.
Bumped version to 0.6.0a1 due to this new feature.

Commit:
less is more

hbldh committed:
Fixes #114 and corrected bug entered by 75f2ff9

Commit:
removed unused

hbldh committed:
Implements scanner for macOS

hbldh committed:
Work on #127 for macOS

hbldh committed:
Implemented first draft of BlueZ Scanner.

synicalsyntax committed:
macOS: Adapt characteristic write log to account for WriteWithoutResponse.

f2aa8d9 fails to account for an edge case the issues found in #114.

mickael9 committed:
backends/dotnet: run callbacks from main thread

Since we're an async library, we probably don't want to run the
callbacks from a different thread than the one the event loop is running
from.

hbldh committed:
Fix for problem created by 838e3a7

hbldh committed:
Fixes #145.

hbldh committed:
Try-Except for probably unnecessary reactor.stop calls

Commit:
corebluetooth. loop is optional

same in other backends

Commit:
separate backend/device

Commit:
bluez: only subscribe to BlueZ messages on DBus

The management of the discovery and clients on Linux is primarly done
through subscribing and listening to DBus messages. The subscriptions
have been identified only through the Interface names and their members.
However, those interfaces are generic and can be implemented by all kind
of applications that expose their API through DBus. This led to the
situation where for example NetworkManager events are forwarded to the
bleak application and handled in the callback.

Although those messages have been ignored, they shouldn't even reach the
application. This patch limits the subscriptions to messages with the
"/org/bluez" path prefix and therefore to relevant BlueZ messages.

Fixes #151

Commit:
bluez: Cleanup all resources even if the client disconnects

Recently there have been changes to make sure we clean up all DBus and
Twisted resources when we disconnect from a server to prevent memory
leaks and running out of available DBus connections. However, this was
only implemented for the case where the application actively disconnects
form the server. If the disconnection was initiated by the server for
example or by going out of range, the dbus connections still leaked.

This patch splits the cleanup into two different methods. One for
cleaning the subscriptions to DBus messages (for example through
notifications) and one for cleaning up the DBus and Twisted resources.
The split is necessary because we want to first remove all subscriptions
before initiating a disconnection and eventually clean up the resources
in the active case, but in the inactive disconnection case we want to
cleanup all resources immediately.

Commit:
Revert "fix(discovery): fix synchronous discover function"

This reverts commit 0f8b05c71a611f8b3b8a03b03e3dcc6b6e209c54.

Commit:
Revert "feat(bluez): add support for asynchronous scanning of devices"

This reverts commit f845869b683152a8d1bd7dfb87e8e3380e34df97.

Commit:
Revert "Do not start a scan if the timeout is <= 0"

This reverts commit 758a14e9929a6aae8d0138fec221bd051c50fd40.

arthur-proglove reviewed (approved):
Looks good to me, once I'm able to run on a device I'll give more feedback :-)

Commit:
fix(client): cache and reuse the reactor based on event loop

The Twisted reactor on POSIX systems leaks file descriptors that are not
cleaned up properly even when the reactor is stopped and presumably
cleaned up. To avoid leaking ressources by creating a new reactor for
every client, we reuse an existing one if the same event loop is used.

Commit:
fix(client): cleanup all ressources if the connection fails

If the connection establishment fails, we have to make sure the
ressources like notifications and dbus/Twisted resources are cleaned up
properly.

Commit:
fix(client): remove all entities of the reactor on client disconnection

Make sure all entities associated with the reactor are cleaned up.
Otherwise ressources like file descriptors leak over multiple
connections.
